---
title: \vspace{-1.5cm} \begin{LARGE} WGS Quality Control Report \end{LARGE}
output:
  pdf_document: 
    latex_engine: lualatex
    keep_tex: true
geometry: margin=0.75in
mainfont: Helvetica
papersize: a4
header-includes:
- \usepackage{titling}
- \pretitle{\begin{flushleft}}
- \posttitle{\end{flushleft}} 
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{float}
- \floatplacement{figure}{H}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{makecell}
- \usepackage{xcolor}
- \usepackage{soul}
- \usepackage{caption}
- \usepackage[singlelinecheck=false]{caption}
- \usepackage[font={small,bf}]{caption}
- \usepackage{multirow}
- \usepackage{array}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \usepackage[dvipsnames]{xcolor}
- \renewcommand{\footnotesize}{\tiny}
- \usepackage{threeparttable}

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.pos = 'H')
```

```{r load library, echo = FALSE, warning = FALSE, message = FALSE}
library(tidyverse)
library(knitr)
library(readxl)
library(dplyr)
library(kableExtra)
library(RColorBrewer)
library(scales)
library(readr)
library(writexl)
library(conflicted)

conflicts_prefer(openxlsx::write.xlsx)
conflicts_prefer(dplyr::filter)
```

```{r load_data, echo = FALSE, warning = FALSE, message = FALSE}
#rename species column to wgs_org
wgs_df <- wgs_df %>% rename(wgs_id = species)

#complete wgs_org name
wgs_df$wgs_id <- ifelse(wgs_df$wgs_id == "Acinetobacter", "Acinetobacter baumannii", 
                         ifelse(wgs_df$wgs_id == "Escherichia", "Escherichia coli", wgs_df$wgs_id))

#load organism code reference file
org_df <- read_xlsx("data_files/whonet_codes_2024.xlsx", "ORGANISM")
org_df <- subset(org_df, select = c(ORG_ARS, ORGANISM_name))
#rename ORGANISM_name column to wgs_id
org_df <- org_df %>% rename(wgs_id = ORGANISM_name)


wgs_df <- merge(wgs_df, org_df, by = "wgs_id", all.x = TRUE)

#rename ORG_ARS column to ORG
wgs_df <- wgs_df %>% rename(org_code = ORG_ARS)



wgs_df$org_code[grepl("Salmonella",wgs_df$wgs_id)] = "sal"
wgs_df$org_code[grepl("Shigella",wgs_df$wgs_id)] = "shi"
wgs_df$org_code[grepl("Vibrio",wgs_df$wgs_id)] = "vic"
wgs_df$org_code[grepl("Klebsiella quasipneumoniae",wgs_df$wgs_id)] = "kpn"
wgs_df$org_code[grepl("Klebsiella variicola",wgs_df$wgs_id)] = "kpn"
wgs_df$org_code[grepl("Raoultella planticola",wgs_df$wgs_id)] = "kpn"




## Create reference data with ±2% tolerance built in
bactopia_genome_size <- data.frame(
  org_code = c("aba", "pae","kpn", "eco","efa", "efm","sau", "spn","hin", "ngo","sal", "sat", "pce", "'shi"),
  genome_size = c(4000000, 6300000, 5400000, 5000000, 3000000, 2900000, 2800000, 2200000, 1800000, 2200000, 4900000, 4700000, 8200000, 4500000)
)

# compute total final coverage
wgs_df <- wgs_df %>%
  left_join(bactopia_genome_size, by = "org_code") %>%
  mutate(
    final_coverage = (qc_original_total_bp/genome_size)
    )



## Create reference data of to check genome size
size_ranges <- data.frame(
  org_code = c("aba", "pae","kpn", "eco","efa", "efm","sau", "spn","hin", "ngo","sal", "sat", "pce", "shi"),
  size_min = c(3400000, 5500000, 5000000, 4500000, 2800000, 2600000,   
             2700000, 2000000, 1700000, 2100000, 4500000, 4600000, 7600000, 4100000),  
  size_max = c(4200000, 7000000, 5700000, 5500000, 3200000, 3100000,   
               3000000, 2300000, 1900000, 2300000, 5300000, 4800000, 8700000, 4800000) 
)

# Modify size ranges: subtract 10% from min, add 10% to max
size_ranges <- size_ranges %>%
  mutate(
    size_min = size_min * 0.9,
    size_max = size_max * 1.1
  )

## Check if GC is within ±2% of expected range
wgs_df <- wgs_df %>%
  left_join(size_ranges, by = "org_code") %>%
  mutate(
    genome_size_result = case_when(
      total_contig_length >= size_min & total_contig_length <= size_max ~ "passed",
      total_contig_length < size_min ~ "failed",
      total_contig_length > size_max ~ "failed"
    )
  ) %>%
  select(org_code, total_contig_length, genome_size_result, everything())




## Create reference data with ±2% tolerance built in
organism_gc_ranges <- data.frame(
  org_code = c("aba", "pae","kpn", "eco","efa", "efm","sau", "spn","hin", "ngo","sal", "sat", "pce", "'shi"),
  GC_min = c(37, 63, 55, 49, 35, 36, 31, 38, 36, 50, 50, 50, 64.7, 48.7),  # Original - 2%
  GC_max = c(41, 69, 59, 53, 39, 40, 35, 42, 40, 54, 54, 54, 68.7, 52.7)   # Original + 2%
)


## Check if GC is within ±2% of expected range
wgs_df <- wgs_df %>%
  left_join(organism_gc_ranges, by = "org_code") %>%
  mutate(
    GC_result = case_when(
      gc_content >= GC_min & gc_content <= GC_max ~ "passed",
      gc_content < GC_min ~ "failed",
      gc_content > GC_max ~ "failed"
    ),
    Original_GC_range = case_when(
      org_code == "pae" ~ "65-67%",
      org_code %in% c("aba", "kpn",
                     "eco", "efa",
                     "efm", "sau",
                     "spn", "hin",
                     "ngo", "sal",
                     "sal") ~ paste0("≈", GC_min + 2, "%")
    )
  ) %>%
  select(org_code, gc_content, Original_GC_range, GC_result, everything())





wgs_df <- wgs_df %>%
  rowwise() %>%
  mutate(
    condition_checks = list(c(
      "Completeness" = completeness >= 90,
      "Contamination" = contamination <= 5,
      "Depth of coverage" = final_coverage >= 20,
      "Genome size" = genome_size_result == "passed",
      "GC content" = GC_result == "passed",
      "Contig count" = total_contig < 500,
      #"" = assembler_l50_contig_count <= 50,
      #"" = assembler_max_contig_length > 200000,
      #"" = assembler_contig_percent_n < 1,
      #"" = assembler_percent_contigs_greater_10k >= 90,
      "N50" = n50_contig_length > 20000,
      "Mean read Q-score" = qc_final_qual_mean >= 30
      
    )),
    
    # Overall result (TRUE if all conditions passed)
    result = all(unlist(condition_checks)),
    
    # List of failed conditions (only if result is FALSE)
    failed_conditions = ifelse(
      result,
      NA_character_,
      paste(names(condition_checks)[!unlist(condition_checks)], collapse = ", ")
    )
  ) %>%
  ungroup()

#add row number
wgs_df$iso_num <- seq.int(nrow(wgs_df))


#file_name <- paste0("data_files/",batch_code,"_wgs_df.xlsx")
#write_xlsx(wgs_df, file_name)

```



```{r sample_sheet, echo = FALSE, warning = FALSE, message = FALSE}
wgs_samplesheet <- read_csv(paste0("E:/DMU Projects/wgs-qc/data_files/",get_samplesheet,".csv"))
names(wgs_samplesheet) <- tolower(names(wgs_samplesheet))

wgs_samplesheet$sample_id <- toupper(wgs_samplesheet$sample_id)


wgs_samplesheet$sample_id <- gsub("-", "_", wgs_samplesheet$sample_id, fixed=TRUE)
#wgs_samplesheet$sample_name <- toupper(wgs_samplesheet$sample_name)


#remove special characters in batch name
batch_name_clean <- gsub("[[:punct:]]", "", batch_code)

# paste batch_name_clean to sample_id column
wgs_samplesheet$sample_id <- paste(wgs_samplesheet$sample_id, batch_name_clean, sep = "_")


#Check if STC sample is present in the id list
stc_sample <- grep("STC", wgs_samplesheet[['sample_id']], value = TRUE)
stc_sample_count <- length(stc_sample)

if (stc_sample_count !=0){
  wgs_samplesheet$sample_id <- gsub("STC", "STC_", wgs_samplesheet$sample_id, fixed=TRUE)
}


#Get experiment name
WGS_exp_name <- unique(wgs_samplesheet[['experiment name']])



#not identified sample
wgs_df$wgs_id[is.na(wgs_df$wgs_id)] <- 'Not Identified'

```


\normalsize Batch Name: `r batch_code`

\normalsize Experiment Name: `r WGS_exp_name`

\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r qc listing, echo = FALSE, warning = FALSE, message = FALSE}
#filter dataframe based on the sample_id present in the WGS dataframe
wgs_samplesheet_good <- wgs_samplesheet[wgs_samplesheet$sample_id %in% sample_name, ]
wgs_samplesheet_good <- subset(wgs_samplesheet_good, select = c('sample_id','description'))



#merge wgs df and samplesheet
wgs_df <- merge(unique(wgs_df), unique(wgs_samplesheet_good), by = "sample_id", all.x= TRUE) 


retain_column <- c('iso_num','sample_id','wgs_id','description')
wgs_result_df <- subset(wgs_df, select = retain_column)

wgs_result_df1 <-wgs_result_df
#wgs_result_df <-wgs_result_df1

#remove batch name in sample_id
wgs_result_df$sample_id <- sub("^(([^_]+_[^_]+)).*", "\\1", wgs_result_df$sample_id)
# Remove trailing underscore
wgs_result_df$sample_id <- sub("_$", "", wgs_result_df$sample_id)



#remove batch name in sample_id
arsrl_result_df$sample_id <- sub("^(([^_]+_[^_]+)).*", "\\1", arsrl_result_df$sample_id)
# Remove trailing underscore
arsrl_result_df$sample_id <- sub("_$", "", arsrl_result_df$sample_id)



#merge ARSRL result and WGS result
qc_listing_df <- merge(unique(arsrl_result_df), unique(wgs_result_df), by = "sample_id", all.x= TRUE) 


#create dataframe ARSRL result with no WGS result
wgs_no_result <- arsrl_result_df[!arsrl_result_df$sample_id %in% wgs_result_df$sample_id, ]



#format wgs_no_result dataframe
if (nrow(wgs_no_result != 0)){
  qc_listing_df$description <- ifelse(is.na(qc_listing_df$description),wgs_samplesheet_good$description[match(paste0(wgs_no_result$sample_id, "_", batch_name_clean), wgs_samplesheet_good$sample_id)], qc_listing_df$description)
  
  qc_listing_df$iso_num <- ifelse(is.na(qc_listing_df$iso_num), nrow(qc_listing_df), qc_listing_df$iso_num)
}



# Define a function to get the part before the first white space
get_value_before_whitespace <- function(x) {
  parts <- strsplit(as.character(x), " ")[[1]]
  return(parts[1])
}

# Get genus of ARSRL org result
qc_listing_df$Genus <- sapply(qc_listing_df$arsrl_org, get_value_before_whitespace)

qc_listing_nc <- qc_listing_df %>%
  rowwise() %>%
  mutate(concordant = 
           grepl(Genus, wgs_id)
         )


#sort wgs df and qc listing df by row number
wgs_df <- wgs_df[order(wgs_df$iso_num, decreasing = FALSE), ]
qc_listing_df <- qc_listing_df[order(qc_listing_df$iso_num, decreasing = FALSE), ]
qc_listing_nc <- qc_listing_nc[order(qc_listing_nc$iso_num, decreasing = FALSE), ]

# set color for failed result
failure_color <- which(wgs_df$result == 'FALSE')


#condition for non-concordant
qc_listing_df$arsrl_org <- ifelse(qc_listing_nc$concordant == TRUE,qc_listing_df$arsrl_org, paste0(qc_listing_df$arsrl_org," (x)"))

wgs_df$concordant <- qc_listing_nc$concordant[match(wgs_df$sample_id, paste0(qc_listing_nc$sample_id, "_", batch_name_clean))]


#remove genus column
qc_listing_df = subset(qc_listing_df, select = -c(Genus) )

#qc_listing_df$sample_id <- gsub("_", "\\_", qc_listing_df$sample_id, fixed = TRUE)


# rename columns
colnames(qc_listing_df) <- c('Sample ID','ARSRL','Isolate No.','WGS','Description')


# arrange columns
qc_listing_df <- qc_listing_df[, c(3,1,5,2,4)] 



qc_listing_df %>% 
  kable(booktabs = T, align = "c", escape = FALSE, row.names = FALSE)%>%
  column_spec(1, width = "1cm") %>%
  column_spec(2, width = "2.8cm") %>%
  column_spec(3, width = "1.5cm") %>%
  column_spec(4, width = "5cm", italic = TRUE) %>%
  column_spec(5, width = "5cm", italic = TRUE) %>%
  row_spec(failure_color, background = "#FD7979") %>%
  row_spec(0, background = "#D4D4D4",bold = TRUE,align = "c") %>%
  footnote(general = c("PASS","  |  ", 
                        "\\\\colorbox{Salmon}{FAILURE}","  |  ",
                        "\\\\textcolor{Blue}{EXCEEDS THRESHOLD METRIC/S}","  |  ",
                        "(x) - NON-CONCORDANT","  |"), 
            general_title = "Legend:",
            footnote_as_chunk = T, escape = F)


```


\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r low reads, echo=FALSE, message=FALSE, warning=FALSE,results='asis'}
#filter dataframe based on the sample_id not present in the WGS samplesheet
wgs_lowreads <- wgs_samplesheet[!(wgs_samplesheet$sample_id %in% sample_name), ]
wgs_lowreads <- subset(wgs_lowreads, select = c('sample_id','description'))

#wgs_lowreads$sample_id <- gsub('_', '\\_', wgs_lowreads$sample_id)
#wgs_lowreads$sample_id <- str_replace_all(wgs_lowreads$sample_id, fixed("_"), "\\_")

if (nrow(wgs_lowreads != 0)){
  lowreads_table <- wgs_lowreads
  lowreads_table$remarks <- c("low read count")
  #lowreads_table$sample_id <- str_replace_all(lowreads_table$sample_id, fixed("_"), "\\_")
  colnames(lowreads_table) <- c('Sample ID','Description','Remarks') 
  
  kable(lowreads_table,  "latex", booktabs = T, align = "c") %>%
    kable_styling(position = "left") %>%
    column_spec(1, width = "3cm") %>%
    column_spec(2, width = "3cm") %>%
    column_spec(3, width = "7cm") %>%
    add_header_above(c("Sample excluded in the analysis" = 3), align = "l", bold = TRUE) %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE,align = "c")
}

```



$\\$
\newpage
\begin{landscape}
\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r summary table, echo = FALSE, warning = FALSE, message = FALSE}

retain_column <- c('iso_num','sample_id','wgs_id','completeness','contamination','final_coverage','total_contig_length','gc_content','total_contig','n50_contig_length','qc_final_qual_mean', 'condition_checks')

wgs_summary_df <- subset(wgs_df, select = retain_column)

# identify the index of columns with false result
wgs_summary_df <- wgs_summary_df %>%
  rowwise() %>%
  mutate(
    # Get column indices of FALSE results
    col_index = ifelse(
      all(unlist(condition_checks)),
      NA_character_,
      paste(which(!unlist(condition_checks)), collapse = ",")
    )
  ) %>%
  select(-condition_checks)


# round off qc_final_coverage column to one decimal place
#wgs_summary_df$qc_final_coverage <-  format(round(wgs_summary_df$qc_final_coverage, 1), nsmall = 1)
wgs_summary_df$final_coverage <- format(round(unlist(wgs_summary_df$final_coverage), 1), nsmall = 1)

# Create a conversion function to convert Basepairs to MEgabasepairs
bp_to_mb <- function(bp) {
  x <- bp / 1e6
  x <- format(round(x, 2), nsmall = 2)
  x <- paste0(x, " Mb")
  return(x)
}

#convert bp to mb
#wgs_summary_df$total_contig_length <- bp_to_mb(wgs_summary_df$total_contig_length)
wgs_summary_df$total_contig_length <- bp_to_mb(unlist(wgs_summary_df$total_contig_length))

summary_table <- function(data) {
  # Convert col_index to character if it isn't already
  data$col_index <- as.character(data$col_index)
  
  # Convert final_coverage to numeric if it isn't already
  data$final_coverage <- as.numeric(data$final_coverage)
  
  #remove batch name in sample_id
  data$sample_id <- sub("^(([^_]+_[^_]+)).*", "\\1", data$sample_id)
  # Remove trailing underscore
  data$sample_id <- sub("_$", "", data$sample_id)
    
  
  data$sample_id <- gsub("_", "\\_", data$sample_id, fixed = TRUE)
  #data$qc_final_qual_mean <- format(round(data$qc_final_qual_mean, 1), nsmall = 1)
  data$qc_final_qual_mean <- format(round(unlist(data$qc_final_qual_mean), 1), nsmall = 1)
  data$final_coverage <- format(round(unlist(data$final_coverage), 1), nsmall = 1)
  
  # Create a display copy with all columns as character
  display_df <<- data %>%
    mutate(across(everything(), as.character)) %>%
    rename(`Isolate No.` = iso_num,
           `Sample ID` = sample_id,
           `WGS ID` = wgs_id,
           `completeness` = completeness,
           `contamination` = contamination,
           `Depth of coverage` = final_coverage,
           `Genome size` = total_contig_length,
           `GC content` = gc_content,
           `Contig count` = total_contig,
           `N50` = n50_contig_length,
           `Mean read Q-score` = qc_final_qual_mean) %>%
    select(-col_index)
  
  
  # Apply formatting
  for (i in 1:nrow(data)) {
    if (!is.na(data$col_index[i]) && nzchar(data$col_index[i])) {
      # Split and convert to numeric safely
      indices <- tryCatch({
        as.numeric(trimws(unlist(strsplit(data$col_index[i], ","))))
      }, error = function(e) numeric(0))
      
      # Remove any NAs from conversion
      indices <- indices[!is.na(indices) & indices > 0]
      
      # Apply formatting to each specified column
      for (col in indices) {
        # Adjust column index (now accounts for renamed columns)
        display_col <- col + 3  # Adjust based on your column structure
        
        if (display_col <= ncol(display_df)) {
          # Get original value
          original_value <- data[i, display_col]  # Adjust offset as needed
          
          # Apply LaTeX formatting for PDF
          display_df[i, display_col] <- paste0("\\textcolor{blue}{\\textbf{", 
                                             as.character(original_value), "}}")
        }
      }
    }
  }
  
  # Create the table
  kable(display_df, format = "latex", booktabs = TRUE, escape = FALSE, align = "c") %>%
    kable_styling(latex_options = c("scale_down", "hold_position")) %>%
    row_spec(failure_color, background = "#FD7979") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
    column_spec(3,italic = TRUE) %>%
    footnote(general = c("PASS","  |  ", 
                        "\\\\colorbox{Salmon}{FAILURE}","  |  ",
                        "\\\\textcolor{Blue}{EXCEEDS THRESHOLD METRIC/S}","  |  ",
                        "(x) - NON-CONCORDANT","  |"), 
            general_title = "Legend:",
            footnote_as_chunk = T, escape = F)
}


summary_table(wgs_summary_df)

```




```{r summary table result, echo = FALSE, warning = FALSE, message = FALSE}
# create new wgs_summary_table with additional column for result
wgs_summary_result <- wgs_df %>%
  select(sample_id, failed_conditions, result) %>%
  mutate(
    batch_code = batch_code,
    result = ifelse(result != TRUE, "Failed", "Passed")
  ) %>%
  relocate(batch_code, .before = sample_id)

# append googlesheet "result" tab
sheet_append(ss = sheet_id, sheet = "result", data = wgs_summary_result)
```




$\\$ $\\$ $\\$ $\color{red}{\normalsize\textbf{RECOMMENDATION:}}$


```{r recommendation table ,echo=FALSE, message=FALSE, warning=FALSE}
wgs_qc <- qc_listing_nc
wgs_qc$failed_conditions = wgs_df$failed_conditions[match(paste0(qc_listing_nc$sample_id, "_", batch_name_clean), wgs_df$sample_id)]

reco_df <- wgs_qc %>%
  filter(!is.na(failed_conditions)|concordant == FALSE) %>%
  select(sample_id, failed_conditions, concordant) %>%
  mutate(
    remarks = ifelse(concordant != TRUE, "Non-concordant result - Check Label", "For repeat testing")
  ) %>%
  select(-concordant) %>%
  group_by(failed_conditions) %>%
  summarise(sample_id = paste(sample_id, collapse = ", "),
            Remarks = paste(unique(remarks), collapse = "; "),
            .groups = 'drop')

# arrange columns
reco_df <- reco_df[, c(2,1,3)] 




if (nrow(wgs_lowreads) != 0){
  wgs_lowreads_reco <- wgs_lowreads %>%
    mutate(
      sample_id = sub("^(([^_]+_[^_]+)).*", "\\1", sample_id),
      sample_id = sub("_$", "", sample_id),
      failed_conditions = "Low read counts",
      remarks = "For repeat testing"
    ) %>%
    select(sample_id, failed_conditions, remarks) %>%
    group_by(failed_conditions) %>%
    summarise(sample_id = paste(sample_id, collapse = ", "),
            Remarks = paste(unique(remarks), collapse = "; "),
            .groups = 'drop')
  
  reco_df <- rbind(reco_df, wgs_lowreads_reco)
}





if (nrow(reco_df) != 0) {
  colnames(reco_df) <- c('Sample ID','Reason - Failed Metrics','Remarks')
  
  reco_table <- reco_df %>% 
  kable(
    format = "latex",
    booktabs = TRUE,
    escape = TRUE,  # Keep FALSE since you pre-escaped underscores
    align = c("l", "c", "c")
  ) %>%
  kable_styling(
    position = "left",
    latex_options = c("striped", "hold_position")
  ) %>%
  column_spec(1, width = "6cm") %>%
  column_spec(2, width = "6cm") %>%
  column_spec(3, width = "4cm") %>%
  row_spec(0, background = "#D4D4D4", bold = TRUE)

  reco_table  
  
}else{
  
  sample_id <- ""
  failed_conditions <- "No further action required for this batch."
  Remarks <- ""
  reco_df <- data.frame(sample_id,failed_conditions, Remarks)
  
  colnames(reco_df) <- c('Sample ID','Reason - Failed Metrics','Remarks')


  reco_table <- reco_df %>% 
  kable(
    format = "latex",
    booktabs = TRUE,
    escape = TRUE,  # Keep FALSE since you pre-escaped underscores
    align = c("l", "c", "c")
  ) %>%
  kable_styling(
    position = "left",
    latex_options = c("striped", "hold_position")
  ) %>%
  column_spec(1, width = "6cm") %>%
  column_spec(2, width = "6cm") %>%
  column_spec(3, width = "4cm") %>%
  row_spec(0, background = "#D4D4D4", bold = TRUE)

reco_table  
}



  

```



\end{landscape}


\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}


```{r qc listing summary narrative, echo=FALSE, message=FALSE, warning=FALSE}
qc_summary_df <- qc_listing_df %>% filter(!is.na(WGS))  %>% count(WGS)

#change column name and sort
names(qc_summary_df)[names(qc_summary_df) == 'n'] <- 'number'
qc_summary_df <- qc_summary_df[order(-qc_summary_df$number),]

# count not identified WGS result
qc_not_identified <- ifelse(qc_summary_df$WGS == 'Not Identified', qc_summary_df$number,0)
qc_not_identified <- unique(qc_not_identified)


qc_tl <- ifelse(qc_not_identified == 0, nrow(qc_listing_df), nrow(qc_listing_df) - qc_not_identified)
qc_distinct <- ifelse(qc_not_identified == 0, nrow(qc_summary_df), nrow(qc_summary_df) - qc_not_identified)



qc_pass <- nrow(wgs_df[wgs_df$result == TRUE, ])
qc_failed <- nrow(wgs_df[wgs_df$result == FALSE, ])

qc_pass_pc <- if (qc_pass != 0) (qc_pass/qc_tl) * 100 else 0
qc_pass_pc <- format(round(qc_pass_pc, 2), nsmall = 2)


qc_failed_pc <- if (qc_failed != 0) (qc_failed/qc_tl) * 100 else 0
qc_failed_pc <- format(round(qc_failed_pc, 2), nsmall = 2)

qc_concordance <- nrow(qc_listing_nc[qc_listing_nc$concordant == TRUE, ])
qc_cn_pc <- if (qc_concordance != 0) (qc_concordance/qc_tl) * 100 else 0
qc_cn_pc <- format(round(qc_cn_pc, 2), nsmall = 2)

```




```{r qc listing summary table, echo=FALSE, message=FALSE, warning=FALSE}
colnames(qc_summary_df) <- c('WGS_ID','Number')

qc_summary_df %>% 
  kable(booktabs = T, align = "lc", row.names = FALSE)%>%
  kable_styling(position = "left") %>%
  column_spec(1, width = "8cm", italic = TRUE) %>%
  row_spec(0, background = "#D4D4D4",bold = TRUE)

```



-   $\color{red}`r qc_distinct`$ distinct species were identified among $\color{red}`r qc_tl`$ isolates.

-   $\color{red}`r qc_pass_pc`$ % (n=`r qc_pass`) of the isolates passed the QC, while $\color{red}`r qc_failed_pc`$ % (n=`r qc_failed`) were tagged with failure.

-   Concordance between ARSRL and WGS species report was $\color{red}`r qc_cn_pc`$ %. $\\$


### GRAPHS

\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r pie_chart, echo=FALSE, message=FALSE, warning=FALSE}
# Calculate the counts for each "wgs_id"
wgs_id_count <- table(wgs_df$wgs_id)

# Create a data frame with "wgs_id" and their counts
wgs_count_df <- data.frame(wgs_id = names(wgs_id_count), Frequency = as.vector(wgs_id_count))

# Sort the data frame by the "Frequency" column in descending order
wgs_count_df <- wgs_count_df[order(-wgs_count_df$Frequency),]

# Now, set the factor levels in reverse order for correct sorting
wgs_count_df$wgs_id <- factor(wgs_count_df$wgs_id, levels = rev(wgs_count_df$wgs_id))


ggplot(data = wgs_count_df, aes(x = " ", y = Frequency, fill = wgs_id)) +
  geom_col(color = "black") +
  theme(legend.text = element_text(size=5)) +
  geom_text(aes(label = Frequency),
            position = position_stack(vjust = 0.5)) +
  xlab("") +
  labs(fill="WGS ID") +
  coord_polar(theta = "y") +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid  = element_blank(), 
        legend.text = element_text(face = "italic"))

```


### Result Classification

```{r organism results, echo=FALSE, message=FALSE, warning=FALSE}
# Define the desired order for the 'result' levels
wgs_df$result_cat <- ifelse(wgs_df$result == TRUE, "PASS", "FAILURE")
desired_order <- c("PASS", "FAILURE")
wgs_df$result_cat <- factor(wgs_df$result_cat, levels = desired_order)

result_colors <- c("FAILURE" = "#f3533a", "PASS" = "#8ad879")

ggplot(data = wgs_df, aes(x = org_code, fill = result_cat)) +
  geom_bar() +
  facet_wrap(~ result_cat) +
  coord_flip() +
  scale_fill_manual(values = result_colors) + # Apply the defined colors
  ylab("Frequency") +
  xlab("Organisms") +
  guides(fill = FALSE) +  # Hide the legend for the 'result' variable
  scale_y_continuous(breaks= pretty_breaks())
  #theme_minimal()
```
 
### Number of contigs

```{r contigs_result,echo=FALSE, message=FALSE, warning=FALSE}
# Convert class
wgs_df$total_contig <- as.numeric(unlist(wgs_df$total_contig))

# Create a new variable for color mapping
wgs_df$status <- ifelse(wgs_df$total_contig < 500, "Passed", "Above threshold")

# Plot
ggplot(data = wgs_df) +
  geom_point(mapping = aes(x = total_contig, y = org_code, color = status)) +
  ylab("Organisms") +
  xlab("Number of Contigs") +
  scale_color_manual(
    values = c("Passed" = "#8ad879", "Above threshold" = "#f3533a"),
    labels = c("Passed", "Above threshold")
  ) +
  labs(color = "Contigs")

```

### N50 Value

```{r n50_result ,echo=FALSE, message=FALSE, warning=FALSE}
# Convert class
wgs_df$n50_contig_length <- as.numeric(unlist(wgs_df$n50_contig_length))

ggplot(data = wgs_df) +
  geom_point(mapping = aes(x = n50_contig_length, y = org_code, color = ifelse(n50_contig_length > 20000 , "#8ad879", "#f3533a"))) +
  ylab("Organisms") +
  xlab("N50 Value") +
  scale_color_manual(values = c("#8ad879", "#f3533a"), labels = c("Passed", "Above threshold")) +
  labs(color = "N50 Result") +
  scale_x_continuous(
  labels = scales::comma_format(big.mark = ',',
                                 decimal.mark = '.'))


```

### Total Length

```{r length_result ,echo=FALSE, message=FALSE, warning=FALSE}
# Convert class
wgs_df$total_contig_length <- as.numeric(unlist(wgs_df$total_contig_length))

ggplot(data = wgs_df) +
  geom_point(mapping = aes(x = total_contig_length, y = org_code, color = ifelse(genome_size_result == "passed" , "#8ad879", "#f3533a"))) +
  ylab("Organisms") +
  xlab("Total length Value") +
  scale_color_manual(values = c("#8ad879", "#f3533a"), labels = c("Passed", "Above threshold")) +
  labs(color = "Total length result") +
  scale_x_continuous(
  labels = scales::comma_format(big.mark = ',',
                                 decimal.mark = '.'))


```



```{r mlst_df ,echo=FALSE, message=FALSE, warning=FALSE}
#rename columns
mlst_df_clean <- df_mlst %>%
  rename("sample_id" = 1,
         "species" = 2,
         "MLST" = 3
         ) %>%
  mutate(
    sample_id = sub(".*/(.*)\\.fna$", "\\1", sample_id),
    sample_id = gsub("-", "_", sample_id)
    ) %>%
  filter(str_detect(sample_id, batch_name_clean))


#drop row without species
mlst_df_clean <- subset(mlst_df_clean, mlst_df_clean$species != '-')

#merge wgs_id from wgs_df to mlst_df to set as species value
wgs_species <- subset(wgs_df, select = c("sample_id", "wgs_id"))
mlst_df_clean <- merge(mlst_df_clean,wgs_species,by = "sample_id")

#remove batch name in sample_id
mlst_df_clean$sample_id <- sub("^(([^_]+_[^_]+)).*", "\\1", mlst_df_clean$sample_id)
# Remove trailing underscore
mlst_df_clean$sample_id <- sub("_$", "", mlst_df_clean$sample_id)



#reorder column
mlst_df_clean <- mlst_df_clean %>% 
  relocate(wgs_id, .after = species)

#identify salmonella enterica group
mlst_df_clean$species <- ifelse(mlst_df_clean$wgs_id == 'Salmonella enterica', 'salmonella_enterica', ifelse(mlst_df_clean$wgs_id == 'Salmonella typhoidal','salmonella_typhoidal', mlst_df_clean$species))


#not identified sample
#mlst_df_clean[is.na(mlst_df_clean)] <- 'Not Identified'


#list unique species in the dataframe
species_list <- unique(mlst_df_clean$species)

mlst_df_list = c()

#create multiple dataframe for each species
for(i in species_list) {
  #retain first four columns
  mlst_info_df <- subset(mlst_df_clean, select = 1:4)
  
  #remove demogs column
  mlst_result_df <- subset(mlst_df_clean, , select = -c(1, 2, 3,4))


  #use first row data as column names in r
  colnames(mlst_result_df)=mlst_result_df[c(1),]

  #remove special characters and number from column header
  colnames(mlst_result_df) <- gsub("\\s*\\(.*", "", colnames(mlst_result_df))


  # Apply gsub to get text inside the parenthesis
  mlst_result_df <- apply(mlst_result_df, 2, function(x) gsub("\\(([^()]*)\\)|.", "\\1", x, perl=T))
  
  #remove rows with na value
  mlst_result_df <- mlst_result_df[ , colSums(is.na(mlst_result_df))==0]

  # Convert back to data frame
  mlst_result_df <- as.data.frame(mlst_result_df)

  #combine the two dataframe
  mlst_result_df <- cbind(mlst_info_df,mlst_result_df)
  
  #assign new name to dataframe
  nam <- paste("df", i, sep = "_")
  assign(nam, mlst_result_df[mlst_result_df$species==i,])
  
  mlst_df_list <- c(mlst_df_list, nam)
}


```

\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1}


### MLST RESULTS

```{r mlst_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#list existing sample_id in the mlst table
not_included <- wgs_df$wgs_id[!(wgs_df$wgs_id %in% mlst_df_clean$wgs_id)]

  
for(i in mlst_df_list) {
  df <- get(i)
  
  df <- subset(df, select = -c(species) )
  names(df)[names(df) == 'wgs_id'] <- 'species'
  
  col_num <- ncol(df)
  
  #kable(df,  "latex", booktabs = T, align = "c") %>%
    #kable_styling(position = "left") %>%
    #row_spec(0, background = "#D4D4D4",bold = TRUE)
  
  mlst_table <- kable(df,align = "c", row.names = FALSE) %>%
    kable_styling(latex_options = c("hold_position","scale_down"), position = "left") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE)%>%
    column_spec(1, width = "3cm") %>%
    column_spec(2, width = "3cm", italic = TRUE) %>%
    column_spec(3:col_num, width = "1cm") %>%
    footnote(general ="(-) Not identified",
           general_title = "Legend: ", 
           footnote_as_chunk = T, title_format = c("italic"))


      
  print(mlst_table)

}

```


```{r no_mlst_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#list existing sample_id in the mlst table
no_mlst <- wgs_df$wgs_id[!(wgs_df$wgs_id %in% mlst_df_clean$wgs_id)]

if (length(no_mlst) !=0){
  sample_id <- wgs_df$sample_id[!(wgs_df$sample_id %in% mlst_df_clean$sample_id)]
  wgs_id <- no_mlst
  species <- "No MLST result"
  
  df <- data.frame(sample_id, wgs_id, species)
  
  no_mlst_table <- kable(df,align = "c", row.names = FALSE) %>%
    kable_styling(position = "left") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE)
   


  print(no_mlst_table)
  
}



```

### MLST RESULTS SUMMARY:

```{r mlst_summary ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
mlst_summary <- data.frame(Species=character(), 
                 MLST=character(), 
                 stringsAsFactors=FALSE) 

for(i in mlst_df_list) {
    df <- get(i)
    
  
    #get unique species value
    species <- unique(df$wgs_id)
    
    #remove from list all value with "quasi"
    species <- species[!species %in% grep("quasi", species, value = T)]

    #Summarize and count MLST results
    mlst_count_df <- df %>% group_by(wgs_id,MLST) %>% summarise(count = sum(!is.na(MLST)))
    mlst_count_df$mlst_count <- paste0(mlst_count_df$MLST, " (n= ", mlst_count_df$count, " )")

    #Remove empty MLST result
    #mlst_count_df  <- subset(mlst_count_df , mlst_count_df $MLST !='-')
    mlst_count_summary  <- subset(mlst_count_df , select = c(wgs_id, mlst_count))
    mlst_summary_df <- mlst_count_summary %>% 
      group_by(wgs_id) %>% 
      mutate(mlst_count = paste0(mlst_count, collapse = ", ")) %>%
      distinct()
    
    #assign new name to dataframe
    #nam <- paste("mlst", i, sep = "_")
    #assign(nam, mlst_summary)
    
    mlst_summary <- rbind(mlst_summary,mlst_summary_df)
    
    #add new data to exisiting dataframe
    #mlst_summary[nrow(mlst_summary) + 1,] = c(species, paste0(mlst_count_df$mlst_count, collapse=", "))
  
}


#Print table
 mlst_summary_table <- kable(mlst_summary,align = "l", row.names = FALSE) %>%
    kable_styling(position = "left") %>%
    column_spec(1, width = "6cm", italic = TRUE) %>%
    column_spec(2, width = "10cm") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
    footnote(general ="(-) Not identified",
           general_title = "Legend: ", 
           footnote_as_chunk = T, title_format = c("italic"))
      
  print(mlst_summary_table)
 
```

\newpage
\begin{landscape}
\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}


\normalsize\textbf{AMR PREDICTION RESULTS}\textbf\normalsize

```{r amr_genes ,echo=FALSE, message=FALSE, warning=FALSE}
df_amrfinderplus1 <- df_amrfinderplus

#checkm2 column names to lower
names(df_amrfinderplus1) <- tolower(names(df_amrfinderplus1))


#Change Name to sample_id
names(df_amrfinderplus1)[names(df_amrfinderplus1) == 'name'] <- 'sample_id'
names(df_amrfinderplus1)[names(df_amrfinderplus1) == 'element symbol'] <- 'element_symbol'

#change "-" to "_" in sample_id column
df_amrfinderplus1$sample_id <- gsub("-", "_", df_amrfinderplus1$sample_id, fixed=TRUE)


#merge wgs_id from wgs_df to mlst_df to set as species value
df_amrfinderplus1 <- merge(df_amrfinderplus1,wgs_species,by = "sample_id")

#remove "quasi" to wgs_id column
df_amrfinderplus1$wgs_id <- gsub("quasi", "", df_amrfinderplus1$wgs_id, fixed=TRUE)



# Filter rows where 'name' contains the word "sample"
df_amrfinderplus1 <- df_amrfinderplus1[grepl(batch_name_clean, df_amrfinderplus1$sample_id), ]

#remove batch name in sample_id
df_amrfinderplus1$sample_id <- sub("^(([^_]+_[^_]+)).*", "\\1", df_amrfinderplus1$sample_id)
# Remove trailing underscore
df_amrfinderplus1$sample_id <- sub("_$", "", df_amrfinderplus1$sample_id)



amr_genes_species <- unique(df_amrfinderplus1$wgs_id)

#vector to hold list of amr gene datables
amr_gene_df_list = c()


for(i in amr_genes_species) {
  df <- subset(df_amrfinderplus1, df_amrfinderplus1$wgs_id == i)
  sample_id <- c(df$sample_id)
  
  #create new dataframe to duplicate df_amrfinderplus1
  amr_genes_df <- df_amrfinderplus1
  
  #change "NA" to " " in subclass column
  amr_genes_df$subclass <- gsub("NA", "", amr_genes_df$subclass, fixed=TRUE)
  
  amr_genes_df$Newcols <- paste(amr_genes_df$type,amr_genes_df$subclass, sep = " ")
  
  #change "/" to "/ " in sample_id column
  amr_genes_df$Newcols <- gsub("/", "/ ", amr_genes_df$Newcols, fixed=TRUE)
  
  #subset dataframe to retain needed columns only
  amr_genes_df <- subset(amr_genes_df, select = c(sample_id,Newcols,element_symbol))
  
  
  #convert multiple rows with same Newcols value to single row
  amr_genes_df <- amr_genes_df %>%
    group_by(Newcols, sample_id) %>%
    summarize(Concat = paste0(element_symbol, collapse = ", "))
  
  #Create new dataframe to summarize the result
  amr_genes_df <- amr_genes_df  %>%
    pivot_wider(names_from = Newcols,
                values_from = Concat)
 #assign new name to dataframe
  nam <- paste("amr", i, sep = "_")
  assign(nam, amr_genes_df[amr_genes_df$sample_id %in% sample_id, ])
  
  amr_gene_df_list <- c(amr_gene_df_list, nam)

}


```


\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r amr_genes_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
for(i in amr_gene_df_list) {
  df <- get(i)
  
  wgs_species_1 <- wgs_species
  
  #remove batch name in sample_id
  wgs_species_1$sample_id <- sub("^(([^_]+_[^_]+)).*", "\\1", wgs_species_1$sample_id)
  # Remove trailing underscore
  wgs_species_1$sample_id <- sub("_$", "", wgs_species_1$sample_id)

  
  # Merge wgs_id from wgs_df to amr_gene to set as species value
  df <- merge(df, wgs_species_1, by = "sample_id")
  
  # Change wgs_id to species
  names(df)[names(df) == 'wgs_id'] <- 'species'
  
  # Reorder column
  df <- df %>%
    relocate(species, .after = sample_id)
  
  # Remove columns with all NA values
  df <- df[, colSums(is.na(df)) < nrow(df)]
  
  # Get unique species value
  species <- unique(df$species)
  species_caption <- paste("\\\\textit{", species, "}", sep = "")
  species_caption <- species_caption[1]
  
  # Subset dataframe to retain needed columns only
  df <- subset(df, select = -c(species))
  
  # Parameters for splitting
  max_rows_per_table <- 6
  max_cols_per_table <- 6  # Maximum columns per table (excluding sample_id)
  
  # Get the column names (excluding sample_id)
  all_cols <- setdiff(colnames(df), "sample_id")
  
  # Calculate number of column groups needed
  num_col_groups <- ceiling(length(all_cols) / max_cols_per_table)
  
  # Split rows first
  num_row_groups <- ceiling(nrow(df) / max_rows_per_table)
  
  for(row_group in 1:num_row_groups) {
    # Determine row range for this group
    start_row <- (row_group-1) * max_rows_per_table + 1
    end_row <- min(row_group * max_rows_per_table, nrow(df))
    df_row_subset <- df[start_row:end_row, , drop = FALSE]
    
    # Now split columns for this row group
    for(col_group in 1:num_col_groups) {
      # Determine column indices for this group
      start_col <- (col_group-1) * max_cols_per_table + 1
      end_col <- min(col_group * max_cols_per_table, length(all_cols))
      cols_to_keep <- c("sample_id", all_cols[start_col:end_col])
      
      df_subset <- df_row_subset[, cols_to_keep, drop = FALSE]
      
      # Create table
      amr_gene_table <- knitr::kable(df_subset, format='latex', booktabs = TRUE, align = "c") %>%
        kable_styling(
          position = "left",
          latex_options = c("HOLD_position", "scale_down"), 
          font_size = 7
        ) %>%
        row_spec(0, background = "#D4D4D4", bold = TRUE) %>%
        column_spec(2:ncol(df_subset), width = "3cm") %>%
        add_header_above(setNames(ncol(df_subset), 
                                paste0(species_caption, 
                                      " (Part ", row_group, ".", col_group, ")")), 
                        bold = TRUE, escape = FALSE, align = "l")
      
      print(amr_gene_table)
      
      # Add vertical space between column groups (but not after last one)
      if(col_group < num_col_groups) cat("\n\n\\vspace{5mm}\n\n")
    }
    
    # Add more vertical space between row groups (but not after last one)
    if(row_group < num_row_groups) cat("\n\n\\vspace{10mm}\n\n")
  }
}
```






```{r no_amr_genes_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#list existing sample_id in the mlst table
no_amr <- wgs_df$wgs_id[!(df_amrfinderplus1$wgs_id %in% wgs_df$wgs_id)]

#remove if with quasi value
no_amr <-no_amr[!grepl('quasi', no_amr)]

if (length(no_amr) !=0){
  sample_id <- wgs_df$sample_id[!(wgs_df$sample_id %in% paste(df_amrfinderplus1$sample_name, batch_name_clean, sep = "_"))]
  wgs_id <- no_amr
  remarks <- "No AMR determinant detected"
  
  df <- data.frame(sample_id, wgs_id, remarks)
  
  no_amr_table <- kable(df,align = "c", row.names = FALSE) %>%
    kable_styling(position = "left") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE)
   


  print(no_amr_table)
  
}

```

\end{landscape}




