---
title: \vspace{-1.5cm} \begin{LARGE} WGS Quality Control Report \end{LARGE}
output:
  pdf_document: 
    latex_engine: lualatex
    keep_tex: true
geometry: margin=0.75in
mainfont: Helvetica
papersize: a4
header-includes:
- \usepackage{titling}
- \pretitle{\begin{flushleft}}
- \posttitle{\end{flushleft}} 
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{float}
- \floatplacement{figure}{H}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{makecell}
- \usepackage{xcolor}
- \usepackage{soul}
- \usepackage{caption}
- \usepackage[singlelinecheck=false]{caption}
- \usepackage[font={small,bf}]{caption}
- \usepackage{multirow}
- \usepackage{array}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \usepackage[dvipsnames]{xcolor}
- \renewcommand{\footnotesize}{\tiny}
- \usepackage{threeparttable}

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.pos = 'H')
```

```{r load library, echo = FALSE, warning = FALSE, message = FALSE}
library(tidyverse)
library(knitr)
library(readxl)
library(dplyr)
library(kableExtra)
library(RColorBrewer)
library(scales)
library(readr)
library(writexl)
library(conflicted)

conflicts_prefer(openxlsx::write.xlsx)
conflicts_prefer(dplyr::filter)
```

```{r load_data, echo = FALSE, warning = FALSE, message = FALSE}
wgs_df1 <- wgs_df
#wgs_df <- wgs_df1

#rename species column to wgs_org
wgs_df <- wgs_df %>% rename(wgs_id = species)

#complete wgs_org name
wgs_df$wgs_id <- ifelse(wgs_df$wgs_id == "Acinetobacter", "Acinetobacter baumannii", 
                         ifelse(wgs_df$wgs_id == "Escherichia", "Escherichia coli", wgs_df$wgs_id))

#load organism code reference file
org_df <- read_xlsx("data_files/whonet_codes_2024.xlsx", "ORGANISM")
org_df <- subset(org_df, select = c(ORG_ARS, ORGANISM_name))
#rename ORGANISM_name column to wgs_id
org_df <- org_df %>% rename(wgs_id = ORGANISM_name)


wgs_df <- merge(wgs_df, org_df, by = "wgs_id", all.x = TRUE)
#rename ORG_ARS column to ORG
wgs_df <- wgs_df %>% rename(org_code = ORG_ARS)



wgs_df$ORG[grepl("Salmonella",wgs_df$wgs_id)] = "sal"
wgs_df$ORG[grepl("Shigella",wgs_df$wgs_id)] = "shi"
wgs_df$ORG[grepl("Vibrio",wgs_df$wgs_id)] = "vic"



## 1. Create reference data of to check genome size
size_ranges <- data.frame(
  org_code = c("aba", "pae","kpn", "eco","efa", "efm","sau", "spn","hin", "ngo","sal", "sat"),
  size_min = c(3400000, 5500000, 5000000, 4500000, 2800000, 2600000,   
             2700000, 2000000, 1700000, 2100000, 4500000, 4600000),  
  size_max = c(4200000, 7000000, 5700000, 5500000, 3200000, 3100000,   
               3000000, 2300000, 1900000, 2300000, 5300000, 4800000) 
)

# Modify size ranges: subtract 10% from min, add 10% to max
size_ranges <- size_ranges %>%
  mutate(
    size_min = size_min * 0.9,
    size_max = size_max * 1.1
  )

## 3. Check if GC is within ±2% of expected range
wgs_df <- wgs_df %>%
  left_join(size_ranges, by = "org_code") %>%
  mutate(
    genome_size_result = case_when(
      assembler_total_contig_length >= size_min & assembler_total_contig_length <= size_max ~ "passed",
      assembler_total_contig_length < size_min ~ "failed",
      assembler_total_contig_length > size_max ~ "failed"
    )
  ) %>%
  select(org_code, assembler_total_contig_length, genome_size_result, everything())




## 1. Create reference data with ±2% tolerance built in
organism_gc_ranges <- data.frame(
  org_code = c("aba", "pae","kpn", "eco","efa", "efm","sau", "spn","hin", "ngo","sal", "sat"),
  GC_min = c(37, 63, 55, 49, 35, 36, 31, 38, 36, 50, 50, 50),  # Original - 2%
  GC_max = c(41, 69, 59, 53, 39, 40, 35, 42, 40, 54, 54, 54)   # Original + 2%
)


## 3. Check if GC is within ±2% of expected range
wgs_df <- wgs_df %>%
  mutate(GC_Content = GC_Content * 100) %>%  # Convert to percentage
  left_join(organism_gc_ranges, by = "org_code") %>%
  mutate(
    GC_result = case_when(
      GC_Content >= GC_min & GC_Content <= GC_max ~ "passed",
      GC_Content < GC_min ~ "failed",
      GC_Content > GC_max ~ "failed"
    ),
    Original_GC_range = case_when(
      org_code == "pae" ~ "65-67%",
      org_code %in% c("aba", "kpn",
                     "eco", "efa",
                     "efm", "sau",
                     "spn", "hin",
                     "ngo", "sal",
                     "sal") ~ paste0("≈", GC_min + 2, "%")
    )
  ) %>%
  select(org_code, GC_Content, Original_GC_range, GC_result, everything())




wgs_df <- wgs_df %>%
  rowwise() %>%
  mutate(
    condition_checks = list(c(
      "Completeness" = Completeness >= 90,
      "Contamination" = Contamination <= 5,
      "Depth of coverage" = qc_final_coverage >= 20,
      "Genome size" = genome_size_result == "passed",
      "GC content" = GC_result == "passed",
      "Contig count" = assembler_total_contig < 500,
      #"" = assembler_l50_contig_count <= 50,
      #"" = assembler_max_contig_length > 200000,
      #"" = assembler_contig_percent_n < 1,
      #"" = assembler_percent_contigs_greater_10k >= 90,
      "N50" = assembler_n50_contig_length > 20000,
      "Mean read Q-score" = qc_final_qual_mean >= 30
      
    )),
    
    # Overall result (TRUE if all conditions passed)
    result = all(unlist(condition_checks)),
    
    # List of failed conditions (only if result is FALSE)
    failed_conditions = ifelse(
      result,
      NA_character_,
      paste(names(condition_checks)[!unlist(condition_checks)], collapse = ", ")
    )
  ) %>%
  ungroup()

#add row number
wgs_df$iso_num <- seq.int(nrow(wgs_df))


#file_name <- paste0("data_files/",get_batchname,"/",get_batchname,"_wgs_df.xlsx")
#write_xlsx(wgs_df, file_name)

```



```{r sample_sheet, echo = FALSE, warning = FALSE, message = FALSE}
wgs_samplesheet <- read_xlsx(paste("data_files/",get_samplesheet,".xlsx", sep=""), sheet = "Modified")
names(wgs_samplesheet) <- tolower(names(wgs_samplesheet))
wgs_samplesheet$sample_id <- toupper(wgs_samplesheet$sample_id)

# Define a function to remove the second underscore if the string doesn't contain the specific text
remove_second_underscore <- function(x) {
  if (!grepl('UTP', x)) {
    x <- gsub("^(.*?_.*?)_(.*)", "\\1\\2", x)
  }
  return(x)
}

# Apply the function to the column of strings
wgs_samplesheet <- wgs_samplesheet %>%
  mutate(sample_id = sapply(sample_id, remove_second_underscore))

#Check if STC sample is present in the id list
stc_sample <- grep("STC", wgs_samplesheet[['sample_id']], value = TRUE)
stc_sample_count <- length(stc_sample)

if (stc_sample_count !=0){
  wgs_samplesheet$sample_id <- gsub("STC", "STC_", wgs_samplesheet$sample_id, fixed=TRUE)
}


#Get experiment name
WGS_exp_name <- unique(wgs_samplesheet[['experiment name']])



#not identified sample
wgs_df$wgs_id[is.na(wgs_df$wgs_id)] <- 'Not Identified'

```


\normalsize Batch Name: `r get_batchname`

\normalsize Experiment Name: `r WGS_exp_name`

\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r qc listing, echo = FALSE, warning = FALSE, message = FALSE}
#filter dataframe based on the sample_id present in the WGS dataframe
wgs_samplesheet_good <- wgs_samplesheet[wgs_samplesheet$sample_id %in% id_list, ]
wgs_samplesheet_good <- subset(wgs_samplesheet_good, select = c('sample_id','description'))

#merge wgs df and samplesheet
wgs_df <- merge(unique(wgs_df), unique(wgs_samplesheet_good), by = "sample_id", all.x= TRUE) 


retain_column <- c('iso_num','sample_id','wgs_id','description')
wgs_result_df <- subset(wgs_df, select = retain_column)

#merge ARSRL result and WGS result
qc_listing_df <- merge(unique(arsrl_result_df), unique(wgs_result_df), by = "sample_id", all.x= TRUE) 


#create dataframe ARSRL result with no WGS result
wgs_no_result <- wgs_result_df[!wgs_result_df$sample_id %in% arsrl_result_df$sample_id, ]



#format wgs_no_result dataframe
if (nrow(wgs_no_result != 0)){
  wgs_no_result$arsrl_org <- NA
  wgs_no_result <- wgs_no_result[, colnames(wgs_no_result)[c(2,5,1,3,4)]]

  qc_listing_df <- rbind(qc_listing_df,wgs_no_result )
  
  new_order <- sapply(id_list, function(x,qc_listing_df){which(qc_listing_df$sample_id == x)}, qc_listing_df=qc_listing_df)
  qc_listing_df <- qc_listing_df[new_order,]

}



# Define a function to get the part before the first white space
get_value_before_whitespace <- function(x) {
  parts <- strsplit(as.character(x), " ")[[1]]
  return(parts[1])
}

# Get genus of ARSRL org result
qc_listing_df$Genus <- sapply(qc_listing_df$arsrl_org, get_value_before_whitespace)

qc_listing_nc <- qc_listing_df %>%
  rowwise() %>%
  mutate(concordant = 
           grepl(Genus, wgs_id)
         )


#sort wgs df and qc listing df by row number
wgs_df <- wgs_df[order(wgs_df$iso_num, decreasing = FALSE), ]
qc_listing_df <- qc_listing_df[order(qc_listing_df$iso_num, decreasing = FALSE), ]
qc_listing_nc <- qc_listing_nc[order(qc_listing_nc$iso_num, decreasing = FALSE), ]

# set color for failed result
failure_color <- which(wgs_df$result == 'FALSE')


#condition for non-concordant
qc_listing_df$arsrl_org <- ifelse(qc_listing_nc$concordant == TRUE,qc_listing_df$arsrl_org, paste0(qc_listing_df$arsrl_org," (x)"))

wgs_df$concordant <- qc_listing_nc$concordant[match(wgs_df$sample_id, qc_listing_nc$sample_id)]


#remove genus column
qc_listing_df = subset(qc_listing_df, select = -c(Genus) )


# rename columns
colnames(qc_listing_df) <- c('Sample ID','ARSRL','Isolate No.','WGS','Description')

# arrange columns
qc_listing_df <- qc_listing_df[, c(3,1,5,2,4)] 



qc_listing_df %>% 
  kable(booktabs = T, align = "c", escape = FALSE, row.names = FALSE)%>%
  column_spec(1, width = "1cm") %>%
  column_spec(2, width = "2.8cm") %>%
  column_spec(3, width = "1.5cm") %>%
  column_spec(4, width = "5cm", italic = TRUE) %>%
  column_spec(5, width = "5cm", italic = TRUE) %>%
  row_spec(failure_color, background = "#FD7979") %>%
  row_spec(0, background = "#D4D4D4",bold = TRUE,align = "c") %>%
  footnote(general = c("PASS","  |  ", 
                        "\\\\colorbox{Salmon}{FAILURE}","  |  ",
                        "\\\\textcolor{Blue}{EXCEEDS THRESHOLD METRIC/S}","  |  ",
                        "(x) - NON-CONCORDANT","  |"), 
            general_title = "Legend:",
            footnote_as_chunk = T, escape = F)


```


\newpage
\begin{landscape}
\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r summary table, echo = FALSE, warning = FALSE, message = FALSE}

retain_column <- c('iso_num','sample_id','wgs_id','Completeness','Contamination','qc_final_coverage','assembler_total_contig_length','GC_Content','assembler_total_contig','assembler_n50_contig_length','qc_final_qual_mean', 'condition_checks')

wgs_summary_df <- subset(wgs_df, select = retain_column)

# identify the index of columns with false result
wgs_summary_df <- wgs_summary_df %>%
  rowwise() %>%
  mutate(
    # Get column indices of FALSE results
    col_index = ifelse(
      all(unlist(condition_checks)),
      NA_character_,
      paste(which(!unlist(condition_checks)), collapse = ",")
    )
  ) %>%
  select(-condition_checks)


# round off qc_final_coverage column to one decimal place
wgs_summary_df$qc_final_coverage <-  format(round(wgs_summary_df$qc_final_coverage, 1), nsmall = 1)


# Create a conversion function to convert Basepairs to MEgabasepairs
bp_to_mb <- function(bp) {
  x <- bp / 1e6
  x <- format(round(x, 2), nsmall = 2)
  x <- paste0(x, " Mb")
  return(x)
}

#convert bp to mb
wgs_summary_df$assembler_total_contig_length <- bp_to_mb(wgs_summary_df$assembler_total_contig_length)

summary_table <- function(data) {
  # Convert col_index to character if it isn't already
  data$col_index <- as.character(data$col_index)
  
  data$sample_id <- gsub("_", "\\_", data$sample_id, fixed = TRUE)
  data$qc_final_qual_mean <- format(round(data$qc_final_qual_mean, 1), nsmall = 1)
  
  # Create a display copy with all columns as character
  display_df <<- data %>%
    mutate(across(everything(), as.character)) %>%
    rename(`Isolate No.` = iso_num,
           `Sample ID` = sample_id,
           `WGS ID` = wgs_id,
           `Completeness` = Completeness,
           `Contamination` = Contamination,
           `Depth of coverage` = qc_final_coverage,
           `Genome size` = assembler_total_contig_length,
           `GC content` = GC_Content,
           `Contig count` = assembler_total_contig,
           `N50` = assembler_n50_contig_length,
           `Mean read Q-score` = qc_final_qual_mean) %>%
    select(-col_index)
  
  
  # Apply formatting
  for (i in 1:nrow(data)) {
    if (!is.na(data$col_index[i]) && nzchar(data$col_index[i])) {
      # Split and convert to numeric safely
      indices <- tryCatch({
        as.numeric(trimws(unlist(strsplit(data$col_index[i], ","))))
      }, error = function(e) numeric(0))
      
      # Remove any NAs from conversion
      indices <- indices[!is.na(indices) & indices > 0]
      
      # Apply formatting to each specified column
      for (col in indices) {
        # Adjust column index (now accounts for renamed columns)
        display_col <- col + 3  # Adjust based on your column structure
        
        if (display_col <= ncol(display_df)) {
          # Get original value
          original_value <- data[i, display_col]  # Adjust offset as needed
          
          # Apply LaTeX formatting for PDF
          display_df[i, display_col] <- paste0("\\textcolor{blue}{\\textbf{", 
                                             as.character(original_value), "}}")
        }
      }
    }
  }
  
  # Create the table
  kable(display_df, format = "latex", booktabs = TRUE, escape = FALSE, align = "c") %>%
    kable_styling(latex_options = c("scale_down", "hold_position")) %>%
    row_spec(failure_color, background = "#FD7979") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
    column_spec(3,italic = TRUE) %>%
    footnote(general = c("PASS","  |  ", 
                        "\\\\colorbox{Salmon}{FAILURE}","  |  ",
                        "\\\\textcolor{Blue}{EXCEEDS THRESHOLD METRIC/S}","  |  ",
                        "(x) - NON-CONCORDANT","  |"), 
            general_title = "Legend:",
            footnote_as_chunk = T, escape = F)
}


summary_table(wgs_summary_df)

```


$\\$ $\\$ $\\$ $\color{red}{\normalsize\textbf{RECOMMENDATION:}}$


```{r recommendation table ,echo=FALSE, message=FALSE, warning=FALSE}
reco_df <- wgs_df %>%
  filter(!is.na(failed_conditions)) %>%
  select(sample_id, failed_conditions, concordant) %>%
  mutate(
    remarks = ifelse(concordant != TRUE, "Non-concordant result", "For repeat testing")
  ) %>%
  select(-concordant) %>%
  group_by(failed_conditions) %>%
  summarise(sample_id = paste(sample_id, collapse = ", "),
            Remarks = paste(unique(remarks), collapse = "; "),
            .groups = 'drop') %>%
  rename(`Sample ID` = sample_id, `Reason - Failed Metrics` = failed_conditions)

# arrange columns
reco_df <- reco_df[, c(2,1,3)] 


if (nrow(reco_df) != 0) {
  reco_table <- reco_df %>% 
  kable(
    format = "latex",
    booktabs = TRUE,
    escape = TRUE,  # Keep FALSE since you pre-escaped underscores
    align = c("l", "c", "c")
  ) %>%
  kable_styling(
    position = "left",
    latex_options = c("striped", "hold_position")
  ) %>%
  column_spec(1, width = "6cm") %>%
  column_spec(2, width = "6cm") %>%
  column_spec(3, width = "4cm") %>%
  row_spec(0, background = "#D4D4D4", bold = TRUE)

  reco_table  
  
}else{
  
  sample_id <- ""
  failed_conditions <- "No further action required for this batch."
  Remarks <- ""
  reco_df <- data.frame(sample_id,failed_conditions, Remarks)
  
  colnames(reco_df) <- c('Sample ID','Reason - Failed Metrics','Remarks')


  reco_table <- reco_df %>% 
  kable(
    format = "latex",
    booktabs = TRUE,
    escape = TRUE,  # Keep FALSE since you pre-escaped underscores
    align = c("l", "c", "c")
  ) %>%
  kable_styling(
    position = "left",
    latex_options = c("striped", "hold_position")
  ) %>%
  column_spec(1, width = "6cm") %>%
  column_spec(2, width = "6cm") %>%
  column_spec(3, width = "4cm") %>%
  row_spec(0, background = "#D4D4D4", bold = TRUE)

reco_table  
}



  

```



\end{landscape}


\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}


```{r qc listing summary narrative, echo=FALSE, message=FALSE, warning=FALSE}
qc_summary_df <- qc_listing_df %>% filter(!is.na(WGS))  %>% count(WGS)

#change column name and sort
names(qc_summary_df)[names(qc_summary_df) == 'n'] <- 'number'
qc_summary_df <- qc_summary_df[order(-qc_summary_df$number),]

# count not identified WGS result
qc_not_identified <- ifelse(qc_summary_df$WGS == 'Not Identified', qc_summary_df$number,0)
qc_not_identified <- unique(qc_not_identified)


qc_tl <- ifelse(qc_not_identified == 0, nrow(qc_listing_df), nrow(qc_listing_df) - qc_not_identified)
qc_distinct <- ifelse(qc_not_identified == 0, nrow(qc_summary_df), nrow(qc_summary_df) - qc_not_identified)



qc_pass <- nrow(wgs_df[wgs_df$result == TRUE, ])
qc_failed <- nrow(wgs_df[wgs_df$result == FALSE, ])

qc_pass_pc <- if (qc_pass != 0) (qc_pass/qc_tl) * 100 else 0
qc_pass_pc <- format(round(qc_pass_pc, 2), nsmall = 2)


qc_failed_pc <- if (qc_failed != 0) (qc_failed/qc_tl) * 100 else 0
qc_failed_pc <- format(round(qc_failed_pc, 2), nsmall = 2)

qc_concordance <- nrow(qc_listing_nc[qc_listing_nc$concordant == TRUE, ])
qc_cn_pc <- if (qc_concordance != 0) (qc_concordance/qc_tl) * 100 else 0
qc_cn_pc <- format(round(qc_cn_pc, 2), nsmall = 2)

```




```{r qc listing summary table, echo=FALSE, message=FALSE, warning=FALSE}
colnames(qc_summary_df) <- c('WGS_ID','Number')

qc_summary_df %>% 
  kable(booktabs = T, align = "lc", row.names = FALSE)%>%
  kable_styling(position = "left") %>%
  column_spec(1, width = "8cm", italic = TRUE) %>%
  row_spec(0, background = "#D4D4D4",bold = TRUE)

```



-   $\color{red}`r qc_distinct`$ distinct species were identified among $\color{red}`r qc_tl`$ isolates.

-   $\color{red}`r qc_pass_pc`$ % (n=`r qc_pass`) of the isolates passed the QC, while $\color{red}`r qc_failed_pc`$ % (n=`r qc_failed`) were tagged with failure.

-   Concordance between ARSRL and WGS species report was $\color{red}`r qc_cn_pc`$ %. $\\$


### GRAPHS

\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r pie_chart, echo=FALSE, message=FALSE, warning=FALSE}
# Calculate the counts for each "wgs_id"
wgs_id_count <- table(wgs_df$wgs_id)

# Create a data frame with "wgs_id" and their counts
wgs_count_df <- data.frame(wgs_id = names(wgs_id_count), Frequency = as.vector(wgs_id_count))

# Sort the data frame by the "Frequency" column in descending order
wgs_count_df <- wgs_count_df[order(-wgs_count_df$Frequency),]

# Now, set the factor levels in reverse order for correct sorting
wgs_count_df$wgs_id <- factor(wgs_count_df$wgs_id, levels = rev(wgs_count_df$wgs_id))


ggplot(data = wgs_count_df, aes(x = " ", y = Frequency, fill = wgs_id)) +
  geom_col(color = "black") +
  theme(legend.text = element_text(size=5)) +
  geom_text(aes(label = Frequency),
            position = position_stack(vjust = 0.5)) +
  xlab("") +
  labs(fill="WGS ID") +
  coord_polar(theta = "y") +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid  = element_blank(), 
        legend.text = element_text(face = "italic"))

```


### Result Classification

```{r organism results, echo=FALSE, message=FALSE, warning=FALSE}
# Define the desired order for the 'result' levels
wgs_df$result_cat <- ifelse(wgs_df$result == TRUE, "PASS", "FAILURE")
desired_order <- c("PASS", "FAILURE")
wgs_df$result_cat <- factor(wgs_df$result_cat, levels = desired_order)

result_colors <- c("FAILURE" = "#f3533a", "PASS" = "#8ad879")

ggplot(data = wgs_df, aes(x = org_code, fill = result_cat)) +
  geom_bar() +
  facet_wrap(~ result_cat) +
  coord_flip() +
  scale_fill_manual(values = result_colors) + # Apply the defined colors
  ylab("Frequency") +
  xlab("Organisms") +
  guides(fill = FALSE) +  # Hide the legend for the 'result' variable
  scale_y_continuous(breaks= pretty_breaks())
  #theme_minimal()
```
 
### Number of contigs

```{r ,echo=FALSE, message=FALSE, warning=FALSE}

ggplot(data = wgs_df) +
  geom_point(mapping = aes(x = assembler_total_contig, y = org_code, color = ifelse(assembler_total_contig < 500, "#8ad879", "#f3533a"))) +
  ylab("Organisms") +
  xlab("Number of Contigs") +
  scale_color_manual(values = c("#8ad879", "#f3533a"), labels = c("Passed", "Above threshold")) +
  labs(color = "Contigs")
  #theme_minimal()
```

### N50 Value

```{r n50_result ,echo=FALSE, message=FALSE, warning=FALSE}
ggplot(data = wgs_df) +
  geom_point(mapping = aes(x = assembler_n50_contig_length, y = org_code, color = ifelse(assembler_n50_contig_length > 20000 , "#8ad879", "#f3533a"))) +
  ylab("Organisms") +
  xlab("N50 Value") +
  scale_color_manual(values = c("#8ad879", "#f3533a"), labels = c("Passed", "Above threshold")) +
  labs(color = "N50 Result") +
  scale_x_continuous(
  labels = scales::comma_format(big.mark = ',',
                                 decimal.mark = '.'))


```

### Total Length

```{r length_result ,echo=FALSE, message=FALSE, warning=FALSE}
ggplot(data = wgs_df) +
  geom_point(mapping = aes(x = assembler_total_contig_length, y = org_code, color = ifelse(genome_size_result == "passed" , "#8ad879", "#f3533a"))) +
  ylab("Organisms") +
  xlab("Total length Value") +
  scale_color_manual(values = c("#8ad879", "#f3533a"), labels = c("Passed", "Above threshold")) +
  labs(color = "Total length result") +
  scale_x_continuous(
  labels = scales::comma_format(big.mark = ',',
                                 decimal.mark = '.'))


```



```{r mlst_df ,echo=FALSE, message=FALSE, warning=FALSE}
#Remain Data column
mlst_df_clean <- mlst_df %>%
  rename("sample_id" = 1,
         "species" = 2,
         "MLST" = 3)

#drop row without species
mlst_df_clean <- subset(mlst_df_clean, mlst_df_clean$species != '-')

#extract text before the first period in sample_id column
mlst_df_clean$sample_id <- sub("\\..*", "", mlst_df_clean$sample_id)

#change "-" to "_" in sample_id column
mlst_df_clean$sample_id <- gsub("-", "_", mlst_df_clean$sample_id, fixed=TRUE)

#merge wgs_id from wgs_df to mlst_df to set as species value
wgs_species <- subset(wgs_df, , select = c("sample_id", "wgs_id"))
mlst_df_clean <- merge(mlst_df_clean,wgs_species,by = "sample_id")

#reorder column
mlst_df_clean <- mlst_df_clean %>% 
  relocate(wgs_id, .after = species)

#identify salmonella enterica group
mlst_df_clean$species <- ifelse(mlst_df_clean$wgs_id == 'Salmonella enterica', 'salmonella_enterica', ifelse(mlst_df_clean$wgs_id == 'Salmonella typhoidal','salmonella_typhoidal', mlst_df_clean$species))


#not identified sample
#mlst_df_clean[is.na(mlst_df_clean)] <- 'Not Identified'


#list unique species in the dataframe
species_list <- unique(mlst_df_clean$species)

mlst_df_list = c()

#create multiple dataframe for each species
for(i in species_list) {
  #retain first four columns
  mlst_info_df <- subset(mlst_df_clean, select = 1:4)
  
  #remove demogs column
  mlst_result_df <- subset(mlst_df_clean, , select = -c(1, 2, 3,4))


  #use first row data as column names in r
  colnames(mlst_result_df)=mlst_result_df[c(1),]

  #remove special characters and number from column header
  colnames(mlst_result_df) <- gsub("\\s*\\(.*", "", colnames(mlst_result_df))


  # Apply gsub to get text inside the parenthesis
  mlst_result_df <- apply(mlst_result_df, 2, function(x) gsub("\\(([^()]*)\\)|.", "\\1", x, perl=T))
  
  #remove rows with na value
  mlst_result_df <- mlst_result_df[ , colSums(is.na(mlst_result_df))==0]

  # Convert back to data frame
  mlst_result_df <- as.data.frame(mlst_result_df)

  #combine the two dataframe
  mlst_result_df <- cbind(mlst_info_df,mlst_result_df)
  
  #assign new name to dataframe
  nam <- paste("df", i, sep = "_")
  assign(nam, mlst_result_df[mlst_result_df$species==i,])
  
  mlst_df_list <- c(mlst_df_list, nam)
}


```

\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1}


### MLST RESULTS

```{r mlst_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#list existing sample_id in the mlst table
not_included <- wgs_df$wgs_id[!(wgs_df$wgs_id %in% mlst_df_clean$wgs_id)]

  
for(i in mlst_df_list) {
  df <- get(i)
  
  df <- subset(df, select = -c(species) )
  names(df)[names(df) == 'wgs_id'] <- 'species'
  
  col_num <- ncol(df)
  
  #kable(df,  "latex", booktabs = T, align = "c") %>%
    #kable_styling(position = "left") %>%
    #row_spec(0, background = "#D4D4D4",bold = TRUE)
  
  mlst_table <- kable(df,align = "c", row.names = FALSE) %>%
    kable_styling(latex_options = c("hold_position","scale_down"), position = "left") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE)%>%
    column_spec(1, width = "3cm") %>%
    column_spec(2, width = "3cm", italic = TRUE) %>%
    column_spec(3:col_num, width = "1cm") %>%
    footnote(general ="(-) Not identified",
           general_title = "Legend: ", 
           footnote_as_chunk = T, title_format = c("italic"))


      
  print(mlst_table)

}

```


```{r no_mlst_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#list existing sample_id in the mlst table
no_mlst <- wgs_df$wgs_id[!(wgs_df$wgs_id %in% mlst_df_clean$wgs_id)]

if (length(no_mlst) !=0){
  sample_id <- wgs_df$sample_id[!(wgs_df$sample_id %in% mlst_df_clean$sample_id)]
  wgs_id <- no_mlst
  species <- "No MLST result"
  
  df <- data.frame(sample_id, wgs_id, species)
  
  no_mlst_table <- kable(df,align = "c", row.names = FALSE) %>%
    kable_styling(position = "left") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE)
   


  print(no_mlst_table)
  
}



```

### MLST RESULTS SUMMARY:

```{r mlst_summary ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
mlst_summary <- data.frame(Species=character(), 
                 MLST=character(), 
                 stringsAsFactors=FALSE) 

for(i in mlst_df_list) {
    df <- get(i)
    
  
    #get unique species value
    species <- unique(df$wgs_id)
    
    #remove from list all value with "quasi"
    species <- species[!species %in% grep("quasi", species, value = T)]

    #Summarize and count MLST results
    mlst_count_df <- df %>% group_by(wgs_id,MLST) %>% summarise(count = sum(!is.na(MLST)))
    mlst_count_df$mlst_count <- paste0(mlst_count_df$MLST, " (n= ", mlst_count_df$count, " )")

    #Remove empty MLST result
    #mlst_count_df  <- subset(mlst_count_df , mlst_count_df $MLST !='-')
    mlst_count_summary  <- subset(mlst_count_df , select = c(wgs_id, mlst_count))
    mlst_summary_df <- mlst_count_summary %>% 
      group_by(wgs_id) %>% 
      mutate(mlst_count = paste0(mlst_count, collapse = ", ")) %>%
      distinct()
    
    #assign new name to dataframe
    #nam <- paste("mlst", i, sep = "_")
    #assign(nam, mlst_summary)
    
    mlst_summary <- rbind(mlst_summary,mlst_summary_df)
    
    #add new data to exisiting dataframe
    #mlst_summary[nrow(mlst_summary) + 1,] = c(species, paste0(mlst_count_df$mlst_count, collapse=", "))
  
}


#Print table
 mlst_summary_table <- kable(mlst_summary,align = "l", row.names = FALSE) %>%
    kable_styling(position = "left") %>%
    column_spec(1, width = "6cm", italic = TRUE) %>%
    column_spec(2, width = "10cm") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
    footnote(general ="(-) Not identified",
           general_title = "Legend: ", 
           footnote_as_chunk = T, title_format = c("italic"))
      
  print(mlst_summary_table)
 
```

\newpage
\begin{landscape}
\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}


\normalsize\textbf{AMR PREDICTION RESULTS}\textbf\normalsize

```{r amr_genes ,echo=FALSE, message=FALSE, warning=FALSE}

#Change Name to sample_id
names(AMR_df)[names(AMR_df) == 'Name'] <- 'sample_id'

#change "-" to "_" in sample_id column
AMR_df$sample_id <- gsub("-", "_", AMR_df$sample_id, fixed=TRUE)

#merge wgs_id from wgs_df to mlst_df to set as species value
AMR_df <- merge(AMR_df,wgs_species,by = "sample_id")

#remove "quasi" to wgs_id column
AMR_df$wgs_id <- gsub("quasi", "", AMR_df$wgs_id, fixed=TRUE)

amr_genes_species <- unique(AMR_df$wgs_id)

#vector to hold list of amr gene datables
amr_gene_df_list = c()


for(i in amr_genes_species) {
  df <- subset(AMR_df, AMR_df$wgs_id == i)
  sample_id <- c(df$sample_id)
  
  #create new dataframe to duplicate AMR_df
  amr_genes_df <- AMR_df
  
  amr_genes_df$Newcols <- paste(amr_genes_df$Type,amr_genes_df$Subtype, sep = " ")
  
  #change "/" to "/ " in sample_id column
  amr_genes_df$Newcols <- gsub("/", "/ ", amr_genes_df$Newcols, fixed=TRUE)
  
  #subset dataframe to retain needed columns only
  amr_genes_df <- subset(amr_genes_df, select = c(sample_id,Newcols,Element.symbol))
  
  
  #convert multiple rows with same Newcols value to single row
  amr_genes_df <- amr_genes_df %>%
    group_by(Newcols, sample_id) %>%
    summarize(Concat = paste0(Element.symbol, collapse = ", "))
  
  #Create new dataframe to summarize the result
  amr_genes_df <- amr_genes_df  %>%
    pivot_wider(names_from = Newcols,
                values_from = Concat)
 #assign new name to dataframe
  nam <- paste("amr", i, sep = "_")
  assign(nam, amr_genes_df[amr_genes_df$sample_id %in% sample_id, ])
  
  amr_gene_df_list <- c(amr_gene_df_list, nam)

}


```


\fontsize{7}{8}
\selectfont
\captionsetup[table]{labelformat=empty}
\renewcommand{\arraystretch}{1.2}

```{r amr_genes_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
for(i in amr_gene_df_list) {
  df <- get(i)
  
  #merge wgs_id from wgs_df to amr_gene to set as species value
  df <- merge(df,wgs_species,by = "sample_id")
  
  #not identified sample
  #wgs_species$wgs_id[is.na(wgs_species$wgs_id)] <- 'Not Identified'
  
  #Change wgs_id to species
  names(df)[names(df) == 'wgs_id'] <- 'species'
  
  #reorder column
  df <- df %>%
    relocate(species, .after = sample_id)
  

  df <- df[,colSums(is.na(df))<nrow(df)]
  
  
  
  #get unique species value
  species <- unique(df$species)

  #remove from list all value with "quasi"
  #species <- species[!species %in% grep("quasi", species, value = T)]
  species_caption <- paste("\\\\textit{",species,"}", sep = "")
  species_caption <-  species_caption[1]
  
  #subset dataframe to retain needed columns only
  df <- subset(df, select = -c(species))
  
  
  if (ncol(df) > 6){
    if (ncol(df) > 12){
      col_count <- ncol(df)
      half_col <- (col_count/2) + 1
      col_split <- half_col + 1 
    
      #select specific columns in index range
      df1 <- df[ , 1:half_col]
      df2 <- df[ , col_split:col_count]
    
      df2 <-cbind(df1[,1, drop=FALSE],df2)
    
    
    
      amr_gene_table1 <- knitr::kable(df1,format='latex',booktabs = T,align = "c")%>%
        kable_styling(latex_options = c("HOLD_position", "scale_down")) %>%
        row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
        column_spec(2:ncol(df1), width = "3cm") %>%
        add_header_above(setNames(ncol(df1),species_caption),bold = TRUE, escape = FALSE, align = "l")
    
      print(amr_gene_table1)
    
    
    
      amr_gene_table2 <- knitr::kable(df2,format='latex',booktabs = T,align = "c")%>%
        kable_styling(latex_options = c("HOLD_position", "scale_down")) %>%
        row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
        column_spec(2:ncol(df2), width = "3cm") %>%
        add_header_above(setNames(ncol(df2),species_caption),bold = TRUE, escape = FALSE, align = "l")
    
      print(amr_gene_table2)
    }else{
      amr_gene_table <- knitr::kable(df,format='latex',booktabs = T,align = "c")%>%
        kable_styling(latex_options = c("HOLD_position", "scale_down")) %>%
        row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
        column_spec(2:ncol(df), width = "3cm") %>%
        add_header_above(setNames(ncol(df),species_caption),bold = TRUE, escape = FALSE, align = "l")
    
      print(amr_gene_table)
        
      }
    
  
  }else{
    amr_gene_table <- knitr::kable(df,format='latex',booktabs = T,align = "c")%>%
      kable_styling(position = "left",latex_options = c("HOLD_position", "scale_down")) %>%
      row_spec(0, background = "#D4D4D4",bold = TRUE) %>%
      #column_spec(2, width = "8cm") %>%
      #column_spec(3:ncol(df), width = "3cm") %>%
      add_header_above(setNames(ncol(df),species_caption),bold = TRUE, escape = FALSE, align = "l")

    print(amr_gene_table)
  }
  
  
}

```






```{r no_amr_genes_table ,echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#list existing sample_id in the mlst table
no_amr <- wgs_df$wgs_id[!(wgs_df$wgs_id %in% AMR_df$wgs_id)]

#remove if with quasi value
no_amr <-no_amr[!grepl('quasi', no_amr)]


if (length(no_amr) !=0){
  sample_id <- wgs_df$sample_id[!(wgs_df$sample_id %in% AMR_df$sample_id)]
  wgs_id <- no_amr
  remarks <- "No AMR determinant detected"
  
  df <- data.frame(sample_id, wgs_id, remarks)
  
  no_amr_table <- kable(df,align = "c", row.names = FALSE) %>%
    kable_styling(position = "left") %>%
    row_spec(0, background = "#D4D4D4",bold = TRUE)
   


  print(no_amr_table)
  
}

```

\end{landscape}




